KAFKA-9273: Extract testShouldAutoShutdownOnIncompleteMetadata from S…

albert02lowis opened the pull request:
…treamTableJoinIntegrationTest into its own test

The main goal is to remove usage of embedded broker (EmbeddedKafkaCluster) in AbstractJoinIntegrationTest and its subclasses.
This is because the tests under this class are no longer using the embedded broker, except for two.
testShouldAutoShutdownOnIncompleteMetadata is one of such tests.
Furthermore, this test does not actually perfom stream-table join; it is testing an edge case of joining with a non-existent topic, so it should be in a separate test.

Testing strategy: run existing unit and integration test

@bbejeck 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

abbccdda reviewed (commented):
@bbejeck Could you take a look to make sure this is the only test which needs to be moving out?

abbccdda commented:
retest this please

albert02lowis committed:
KAFKA-9273: Extract testShouldAutoShutdownOnIncompleteMetadata from StreamTableJoinIntegrationTest into its own test

The main goal is to remove usage of embedded broker (EmbeddedKafkaCluster) in AbstractJoinIntegrationTest and its subclasses.
This is because the tests under this class are no longer using the embedded broker, except for two.
testShouldAutoShutdownOnIncompleteMetadata is one of such tests.
Furthermore, this test does not actually perfom stream-table join; it is testing an edge case of joining with a non-existent topic, so it should be in a separate test.

albert02lowis commented:
Hi, there's another unit test that need to be moved out (`StreamStreamJoinIntegrationTest.shouldNotAccessJoinStoresWhenGivingName`) but I thought of doing that in another PR (to make the PRs small). But let me know if it's ok to just include it in this PR.

(Btw, I force-push updated the commit with the right user email)

abbccdda commented:
@albert02lowis I see, let's try to do all of them in one PR then.

albert02lowis committed:
KAFKA-9273: Extract shouldNotAccessJoinStoresWhenGivingName from StreamStreamJoinIntegrationTest into its own test.

shouldNotAccessJoinStoresWhenGivingName is the last unit test that has a dependency to embedded broker in AbstractJoinIntegrationTest.
This test does not actually test the behaviour of stream-stream join;
It is testing an edge case of accessing the store of a join, so it should be in a separate test.

albert02lowis committed:
KAFKA-9273: Remove dependency on embedded broker from AbstractJoinIntegrationTest.

The tests under AbstractJoinIntegrationTest previously requires embedded broker, but it is better to use TopologyTestDriver instead.
Existing stream-stream join tests that still make use of embedded broker is moved to use TopologyTestDriver, by specifying it under BOOTSTRAP_SERVERS_CONFIG.
And then the old EmbeddedKafkaCluster in AbstractJoinIntegrationTest is safely removed.

albert02lowis commented:
Hi @abbccdda sure thing, I have added 2 more commits to this PR and the unit + integration tests also looks good in my local

bbejeck commented:
Ok to test

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/AbstractJoinIntegrationTest.java:
```diff
@@ -123,25 +117,18 @@
     public static void setupConfigsAndUtils() {
 
         STREAMS_CONFIG.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
-        STREAMS_CONFIG.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
```
@albert02lowis I've confirmed locally that the test failures are related.  The `TopologyTestDriver` still needs a bootstrap servers config value.
Adding something like `STREAMS_CONFIG.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "test:0000");` into the `AbstractJoinIntegrationTest.setupConfigsAndUtils()` gets all tests passing. 

I'm not sure how the JDK 14 tests passed as locally I got the same errors as the JDK 8 and 11 build locally.

albert02lowis replied:
I see, let me add back this BOOTSTRAP_SERVERS_CONFIG inside setupConfigsAndUtils then 👍🏻

albert02lowis replied:
Hi @bbejeck I have done the required change in a new commit. I have also extracted out redundant calls to set `BOOTSTRAP_SERVERS_CONFIG` in the subclasses' test methods.

bbejeck reviewed (commented):
@albert02lowis thanks for the contribution.  Overall the PR looks good to me, after the one comment is addressed we'll probably get this merged.

albert02lowis committed:
Centralize setting of BOOTSTRAP_SERVERS_CONFIG inside AbstractJoinIntegrationTest.

Previously the setting of BOOTSTRAP_SERVERS_CONFIG is added on each test method, but this can actually be extracted to the BeforeClass method which is setupConfigsAndUtils in AbstractJoinIntegrationTest.

bbejeck commented:
retest this please

bbejeck commented:
Retest this please.

bbejeck commented:
Ok to test

abbccdda commented:
retest this please

bbejeck commented:
ok to test

bbejeck reviewed (commented):
Thanks for the update @albert02lowis, this LGTM.  The streams build passed for me locally.  Once we can get a build running here, I'll merge it.

bbejeck commented:
retest this please

bbejeck commented:
Java 8 failed on an unrelated test `kafka.api.SaslSslConsumerTest.testCoordinatorFailover` created a Jira ticket.

retest this please

bbejeck commented:
Java 8 passed
Java 14 failed with `org.apache.kafka.streams.integration.PurgeRepartitionTopicIntegrationTest.shouldRestoreState`
Java 11 failed with `kafka.api.PlaintextAdminIntegrationTest.testAlterReplicaLogDirs`

retest this please

bbejeck commented:
Java 11 passed
Java 14 failed due to ` Execution failed for task ':connect:api:javadoc'.` ? (0 tests failed
Java 8 appears hung
Trying this one more time

retest this please

bbejeck commented:
merged #9108 into trunk

Thanks for the contribution @albert02lowis!

albert02lowis commented:
Thanks for the reviews too @bbejeck and @abbccdda ! Really excited as this is my first contribution 🙏🏻
